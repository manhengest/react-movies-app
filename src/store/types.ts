export const MOVIES_TYPES = {
    LOAD_MOVIES: "LOAD_MOVIES",
    UPDATE_MOVIES_COUNT: "UPDATE_MOVIES_COUNT",
    UPDATE_SELECTED_GENRES: "UPDATE_SELECTED_GENRES",
    UPDATE_SELECTED_SORTING: "UPDATE_SELECTED_SORTING",
}

export const APP_TYPES = {
    TOGGLE_MODAL: "TOGGLE_MODAL",
    SET_VIEW_MODE: "SET_VIEW_MODE",
    SET_DETAILED_VIEW_DATA: "SET_DETAILED_VIEW_DATA",
}
