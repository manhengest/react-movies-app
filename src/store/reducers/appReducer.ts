import { AnyAction } from "redux";
import { HYDRATE } from "next-redux-wrapper";
import { createSelector } from "reselect";

import { APP_TYPES } from "../types";
import { UniqMovie } from "../../components/MovieCard/interface";
import { RootState } from "./rootReducer";

export const initialState = {
    isAddModalOpened: false,
    isEditModalOpened: false,
    isDeleteModalOpened: false,
    isSuccessModalOpened: false,
    movieId: null
}

export default function appReducer(state: typeof initialState = initialState, action: AnyAction) {
    switch (action.type) {
        case HYDRATE:
            return {...state, ...action.payload.appData};
        case APP_TYPES.TOGGLE_MODAL: {
            return {
                ...state,
                [action.payload.variable]: action.payload.status,
                movieId: action.payload.movieId
            }
        }
        case APP_TYPES.SET_VIEW_MODE: {
            return {
                ...state,
                viewMode: action.payload
            }
        }
        case APP_TYPES.SET_DETAILED_VIEW_DATA: {
            return {
                ...state,
                detailedViewData: action.payload
            }
        }
        default:
            return state
    }
}

export const toggleModal = (variable: string, status: boolean, movieId?: number) => ({
    type: APP_TYPES.TOGGLE_MODAL,
    payload: {
        variable,
        status,
        movieId
    }
})

export const setViewMode = (mode: string) => ({
    type: APP_TYPES.SET_VIEW_MODE,
    payload: mode
})

export const setDetailedViewData = (data: UniqMovie) => ({
    type: APP_TYPES.SET_DETAILED_VIEW_DATA,
    payload: data
})

export const appStateSelector = (state: RootState) => state.appData;

export const movieIDSelector = createSelector(
    appStateSelector,
    (state) => state.movieId
);
