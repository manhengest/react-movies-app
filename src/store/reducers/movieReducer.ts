import { MOVIES_TYPES } from "../types";
import axios from "axios";
import { Movie, MoviesResponse, ReduxAction } from "../../components/MovieCard/interface";

const initialState = {
    movies: [],
    total: 0,
    selectedGenres: [],
    selectedSorting: "release_date"
}

export default function movieReducer(state: typeof initialState = initialState, action: ReduxAction) {
    switch (action.type) {
        case MOVIES_TYPES.LOAD_MOVIES: {
            return {
                ...state,
                movies: [
                    ...action.payload
                ]
            }
        }
        case MOVIES_TYPES.UPDATE_MOVIES_COUNT: {
            return {
                ...state,
                total: action.payload
            }
        }
        case MOVIES_TYPES.UPDATE_SELECTED_GENRES: {
            return {
                ...state,
                selectedGenres: [
                    ...action.payload
                ]
            }
        }
        case MOVIES_TYPES.UPDATE_SELECTED_SORTING: {
            return {
                ...state,
                selectedSorting: action.payload
            }
        }
        default:
            return state
    }
}

export const updateMovies = (movies: Movie[]) => ({
    type: MOVIES_TYPES.LOAD_MOVIES,
    payload: movies
})

export const updateMoviesCount = (count: number) => ({
    type: MOVIES_TYPES.UPDATE_MOVIES_COUNT,
    payload: count
})

export const updateSelectedGenres = (genres: string[]) => ({
    type: MOVIES_TYPES.UPDATE_SELECTED_GENRES,
    payload: genres
})

export const updateSelectedSorting = (title: string) => ({
    type: MOVIES_TYPES.UPDATE_SELECTED_SORTING,
    payload: title
})

export const fetchMovies = (limit: number = 9) => async (dispatch: any, getState: any) => {
    const { selectedGenres } = getState().moviesData;
    const { selectedSorting } = getState().moviesData;

    const filterParams = {
        searchBy: "",
        sortBy: "",
        filter: "",
        sortOrder: "desc",
        limit
    }

    if (selectedGenres.length) {
        filterParams.searchBy = "genres"
        filterParams.filter = selectedGenres.join(",")
    }

    filterParams.sortBy = selectedSorting

    const response: MoviesResponse = await axios.get("http://localhost:4000/movies", {
        params: filterParams
    })

    dispatch(updateMovies(response.data.data))
    dispatch(updateMoviesCount(response.data.totalAmount))
}

export const createMovie = (data: Movie) => async () => {
    return await axios.post("http://localhost:4000/movies", data)
}

export const getMovie = (id: number) => async () => {
    return await axios.get(`http://localhost:4000/movies/${ id }`)
}

export const updateMovie = (data: Movie) => async () => {
    return await axios.put("http://localhost:4000/movies", data)
}

export const deleteMovie = (id: number) => async () => {
    return await axios.delete(`http://localhost:4000/movies/${ id }`)
}
