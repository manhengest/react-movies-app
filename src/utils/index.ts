import { Genre } from "../components/MovieCard/interface";

export const genres = Object
    .values(Genre)
    .map((title, _id) => ({
        title,
        id: _id + 1,
        active: !_id
    }))

export const movieDurationConverter = (duration: number): string => {
    const hours = (duration / 60);
    const clearedHours = Math.floor(hours);
    const minutes = (hours - clearedHours) * 60;
    const clearedMinutes = Math.round(minutes);
    return clearedMinutes ? `${clearedHours ? clearedHours + "h " : ""}${clearedMinutes}min` : `${clearedHours}h`
}

export const addOrRemoveFromArray = (array: any[], newItem: any) => {
    array.indexOf(newItem) === -1 ? array.push(newItem) : array.splice(array.indexOf(newItem), 1)

    return array
}
