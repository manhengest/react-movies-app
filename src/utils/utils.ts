export const movieDurationConverter = (duration: number): string => {
    const hours = (duration / 60);
    const clearedHours = Math.floor(hours);
    const minutes = (hours - clearedHours) * 60;
    const clearedMinutes = Math.round(minutes);
    return clearedMinutes ? `${clearedHours}h ${clearedMinutes}min` : `${clearedHours}h`
}

export const addOrRemoveFromArray = (array: any[], newItem: any) => {
    array.indexOf(newItem) === -1 ? array.push(newItem) : array.splice(array.indexOf(newItem), 1)

    return array
}
