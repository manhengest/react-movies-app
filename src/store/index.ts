import { createStore, applyMiddleware, Store } from 'redux'
import {createWrapper, Context} from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk'
import { composeWithDevTools } from 'redux-devtools-extension'
import { rootReducer, RootState } from './reducers/rootReducer'

const composedEnhancer = composeWithDevTools(
    applyMiddleware(thunkMiddleware)
)

const makeStore = (context: Context) => createStore(rootReducer, composedEnhancer);
export const wrapper = createWrapper<Store<RootState>>(makeStore, {debug: true});
