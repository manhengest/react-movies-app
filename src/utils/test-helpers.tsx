import React, { ReactChild } from "react";
import { createStore } from "redux";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";

import { rootReducer } from "../store/reducers/rootReducer";

export const renderWithRedux = (
    component: ReactChild,
    { initialState, store = createStore(rootReducer, initialState) } = {}
) => {
    return {
        ...render(<Provider store={ store }>{ component }</Provider>),
        store
    }
}
