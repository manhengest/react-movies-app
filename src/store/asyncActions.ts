import { MoviesResponse } from "../components/MovieCard/interface";
import { updateMovies, updateMoviesCount } from "./reducers/movieReducer";
import { getMovies } from "../api/movies";
import { AxiosResponse } from "axios";
import { ThunkDispatch } from "redux-thunk";
import { Action } from "redux";
import { RootState } from "./index";

export const fetchMovies = (query?: string) => {
    const urlParams = new URLSearchParams(location.search);
    const genre = urlParams.get('genre');
    const filterParams = {
        search: "",
        searchBy: "",
        sortBy: "",
        filter: "",
        sortOrder: "desc",
        limit: 9
    }
    if (query) {
        filterParams.searchBy = "title"
        filterParams.search = query
    }
    if (!query && genre) {
        filterParams.searchBy = "genres"
    }
    if (genre) {
        filterParams.filter = genre
    }

    return (dispatch: ThunkDispatch<RootState, void, Action>) => {
        getMovies(filterParams).then((response: AxiosResponse<MoviesResponse>) => {
            dispatch(updateMovies(response.data.data))
            dispatch(updateMoviesCount(response.data.totalAmount))
        })
    }
}

