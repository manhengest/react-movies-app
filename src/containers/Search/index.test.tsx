import React from "react";
import { render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event";
import configureMockStore from 'redux-mock-store'
import { createMemoryHistory } from "history";
import { Provider } from 'react-redux'
import { Router } from "react-router-dom";
import thunk from 'redux-thunk'
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';

import { Search } from "./index";
import { initialState } from "../../store/reducers/movieReducer";
import { MOVIES_TYPES } from "../../store/types";
import { renderWithRedux } from "../../utils/test-helpers";
import { fetchMovies } from "../../store/asyncActions";

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)
const store = mockStore(initialState)
const mock = new MockAdapter(axios);

describe("Search", () => {
    beforeEach(() => {
        store.clearActions();
    });

    test("load movies", () => {
        mock.onGet("http://localhost:4000/movies").reply(200, {
            data: [
                {
                    id: 1,
                    poster_path: "/test/path",
                    title: "test title",
                    genres: ["test"],
                    release_date: "test",
                    runtime: 20,
                    overview: "test",
                    vote_average: 5
                }
            ],
            totalAmount: 1
        });

        store.dispatch(fetchMovies()).then(() => {
            const expectedActions = [
                {
                    type: MOVIES_TYPES.LOAD_MOVIES,
                    payload: [
                        {
                            id: 1,
                            poster_path: "/test/path",
                            title: "test title",
                            genres: ["test"],
                            release_date: "test",
                            runtime: 20,
                            overview: "test",
                            vote_average: 5
                        }
                    ]
                },
                {
                    type: MOVIES_TYPES.UPDATE_MOVIES_COUNT,
                    payload: 1
                }
            ]

            expect(store.getActions()).toEqual(expectedActions);
        });
    });
    test("render Search snapshot", () => {
        const { asFragment } = renderWithRedux(<Search/>);
        expect(asFragment()).toMatchSnapshot();
    })
    test("render important elements", () => {
        renderWithRedux(<Search/>)

        expect(screen.getByRole("textbox")).toBeInTheDocument();
        expect(screen.getByRole("button")).toBeInTheDocument();
    })
    test("search field events", () => {
        const history = createMemoryHistory()

        render(
            <Router history={ history }>
                <Provider store={ store }>
                    <Search/>
                </Provider>
            </Router>
        )

        userEvent.type(screen.getByRole("textbox"), "Lord")
        expect(screen.getByRole("textbox")).toHaveValue("Lord")

        userEvent.click(screen.getByRole("button"))
        expect(history.location.pathname).toContain("/search/Lord")
        expect(screen.getByRole("textbox")).toHaveValue("")
    })
})
