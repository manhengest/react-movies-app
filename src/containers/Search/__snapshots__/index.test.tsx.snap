// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search render Search snapshot 1`] = `
<DocumentFragment>
  <div
    class="search"
  >
    <span
      class="primary-title primary-title_search"
    >
      FIND YOUR MOViE
    </span>
    <div
      class="search__inner"
    >
      <input
        class="search-field"
        placeholder="What do you want to watch?"
        type="text"
        value=""
      />
      <button
        class="btn primary-button search__btn"
        type="button"
      >
        Search
      </button>
    </div>
  </div>
</DocumentFragment>
`;
